T19-Existing-Items

Digital-Brendan opened the pull request:
Hi all,

For this ticket, I implemented:

- showing all created items to the user
- creating a categoryItem from an existing item
- deleting an item from an existing categoryItem
- deleting all instances of an item (all categoryItems created for it)
- making sure the user knows the consequences of deleting an item
- a new 'categoryItemExists' function.

Basically, to add an existing item to a category, the user can choose to see a list of already existing items and tap on the one they want to add. If it is already present in the category, no duplicate will be created. 

To delete an item, the user can select to 'Delete Everywhere' when tapping on the delete icon beside a categoryItem (this is the only way to truly delete an item). This deletes all categoryItems created for the item, the item's image, then the item itself, as well as updating the ranks of the other categoryItems within the affected categories. The user is warned of the consequences of this before continuing, with traffic-light-coloured buttons indicating severity of deletion.

To the best of my knowledge, I have tested each of these features thoroughly, but of course, if you think of anything else, let me know!

Thanks :)

Commit:
Allow user to add existing items to a category

Commit:
Merge with T10

Commit:
Show option to delete item from all other categories

Commit:
Merge branch 'master' into T19-Existing-Items

Commit:
Remove all categoryItems for a given deleted item

Commit:
Rewrite deleteAllCategoryItemsForItem function

Commit:
Reorder sequence of delete from everywhere functions

Commit:
Test that UI elements are present

Commit:
Merge branch 'master' into T19-Existing-Items

Commit:
Fix broken test

Commit:
Fix overflow error

Commit:
Implement all T19 tests

Commit:
Expand firebase functions test coverage

Commit:
Test that deleting all categoryitems for a given item correctly updates ranks

Commit:
Improve comments

Commit:
Merge branch 'master' into T19-Existing-Items

Commit:
Resolve merge conflicts with T20

Digital-Brendan commented:
**Update**: merge conflicts should have been resolved, and my code is now fully integrated with T20.

Digital-Brendan commented:
Also please note that although the number of line changes seems excessive, the majority of these are due to the automatic reformatting that occurs when saving a file.

absirgue reviewed (changes requested):
Hey,

Thanks for the work!

First little comment, this is not your fault at all I fully understand but would be nice if you can try to remove your code format or something to have the code presented like it is for there rest of us bcs it looks really dense and compact now. (if you can't do it we can find a solution like I could go back and save the files myself for them to be formatted properly before merging but I'm not sure that would work and it's not ideal).

Apart from that, I think there are a few points that must be addressed:
- First, in terms of UI/UX, it is strange to have the widget to add an item look like that. The "create new item" and "use existing item" buttons should not be on the same level and the "use existing item" one should not be able to be seen only after entering a name and an image if you see what I mean. I believe there are 2 solutions for that:
                  1. put the "use existing item" button at the top of this widget and then have the whole section about if you want to create a new item (so the thing to select an image, the thing to enter a name, and the button "create new item" underneath). If that makes any sense.
                  2. make it so that when you add an item, it opens a widget that allows you to choose between using an existing one or creating a new one and redirects you to the page you built or to the widget Ali built depending on the choice.

=> I believe the first one is quite easy to do and may actually look the best. Please lmk if I didn't explain it clearly.

Second UI problem, Can you show the name of the items in addition to the image on the page where I can select which of the existing item I want to add? (makes it clear +_ if I have two pictures for the same item for some reason I can know which is which and which word will be read to the child).

Third, can you add a search bar to this page (it will get loaded with a lot of items)? Yousef did it in his branch I believe so you can check it out but I think it is core to keeping the functionality you just implemented usable.

And yep fourth thing. I love the way you implemented the deleting functionality I didn't notice it at first sight bcs wasn't expecting it to be implemented that way but it's great!! Only thing, once again just a small UX concern, can you make the "delete everywhere" button not stand out that much (less bright red) and make it the middle or leftmost button instead of the rightmost bcs I think as a user you have a tendency to just click on the red button that's the rightmost and that would be bad. On top of that, can you make just a small alert dialog that asks the user to confirm the deletion of ALL the item and category items only because it would be quite painful for them to do it accidentally.

Thanks, let me know if you have any questions :)

Commit:
Change design for adding existing item

Commit:
Implement Anton's changes about delete confirmation button colour and additional deletion confirmation dialogue

Commit:
Remove unneccessary test

Commit:
Add test.ensureVisible to relevant add item tests

Commit:
Fix pod file

Commit:
Fix pod file again

Commit:
Pod file error

Commit:
Run pod install

Commit:
Update podfile.lock

Commit:
Undo changes to add existing item

Commit:
Add search bar to existing items page

Commit:
Add test for searchbar

absirgue reviewed (approved):
Thanks for integrating all the requested changes. Looks perfect to me!

Ali-Alkhars commented on lib/widgets/existing_items_grid.dart:
```diff
@@ -0,0 +1,133 @@
+  }
+
+  void addItemAsCategoryItem({required String itemId, required String imageUrl, required String name, required BuildContext context}) async {
+    DocumentSnapshot categoryItem = await widget.firestore.collection('categoryItems/${widget.categoryId}/items').doc(itemId).get();
```
move this to a new method in firebase_functions

Digital-Brendan replied:
I'm not sure how to do this since it needs the context of its parent widget? (where to redirect user and what snackbar message to display)

Ali-Alkhars replied:
I think you misunderstood me here. I meant move only `DocumentSnapshot categoryItem = await widget.firestore.collection('categoryItems/${widget.categoryId}/items').doc(itemId).get();` to `firebase_functions` and call it here

Digital-Brendan replied:
Oh that makes much more sense

Ali-Alkhars commented on lib/pages/admin/add_existing_item.dart:
```diff
@@ -0,0 +1,64 @@
+          key: Key('app_bar'),
+          title: const Text('Add an Existing Item'),
+        ),
+        drawer: const AdminSideMenu(),
```
Would be better for UX if you change the menu drawer to a back button

Digital-Brendan replied:
This is good though I like the idea of being able to use Android's native back button and still have access to the admin side menu without exiting the current page. I'm not sure which is a better UX?

Ali-Alkhars replied:
I see, but there’s no back button for ios, so I think it’s better to make a back button instead of a menu.

Ali-Alkhars commented on lib/pages/admin/add_existing_item.dart:
```diff
@@ -0,0 +1,64 @@
+        } else if (snapshot.hasError) {
+          return Text('Error: ${snapshot.error}');
+        } else {
+          return CircularProgressIndicator();
```
Use the `custom_loading_indicator.dart` widget instead of this. It looks way better

Ali-Alkhars commented on lib/pages/admin/add_choice_board_item.dart:
```diff
@@ -61,22 +52,35 @@ class _AddChoiceBoardItem extends State<AddChoiceBoardItem> {
                   mainAxisSize: MainAxisSize.min,
                   // page (Hero) contents
                   children: [
+                    Padding(padding: const EdgeInsets.fromLTRB(16, 16, 16, 8), child: Text("Already have an item in mind?"),),
```
Hardcode the text colour to black. In the begging, I had the text colour white and I didn't notice it.

Also, increase the text size? Maybe the same size as the "Pick an image" text if it looks good

Ali-Alkhars commented on lib/widgets/existing_items_grid.dart:
```diff
@@ -0,0 +1,133 @@
+    firestoreFunctions = FirebaseFunctions(auth: this.auth, firestore: this.firestore, storage: this.storage);
+  }
+
+  final List<ImageDetails> imagesList;
```
move this to the top with the other fields

Ali-Alkhars commented on lib/widgets/existing_items_grid.dart:
```diff
@@ -0,0 +1,133 @@
+  @override
+  Widget build(BuildContext context) {
+    if (widget.imagesList.isEmpty) {
+      return Text("No items to show. Add some in the 'Choice Board' page");
```
Wrap this with a `Center` and increase the text size so it looks nicer

Ali-Alkhars reviewed (changes requested):
Hey Brendan, really great work!

I added some comments below but they are all minor stuff to change, so should not take much time.

**There is also this little bug:** Adding a pre-existing item should take the current availability of the item, not become True on default.

Commit:
implement Alis styling changes

Commit:
Create getItemAvailability function

Digital-Brendan commented:
Hey @Ali-Alkhars that should be everything implemented! Thanks

Commit:
Add back button

Commit:
Add leading back button

skellyhoe committed:
Add back button.

skellyhoe committed:
Delete comment.

Commit:
Revert "Delete comment."

This reverts commit 8bfd407379b73a63f0968f3325dbc75f71e8a1e1.

Commit:
Revert "Fix add existing item test"

This reverts commit db7de981e87bae319c92b70e6fe0a9a6f294d001.

Commit:
Fix

Commit:
Revert "Create getItemAvailability function"

This reverts commit a30b0a79c49395ff00a522e37304591bf036fe0a.

skellyhoe committed:
test 2

skellyhoe committed:
Attempt fix merge conflicts.

skellyhoe commented:
I tried to fix conflicts I think it should be good now. There's one test that fails and I think its because the delete item button isn't doing pushAndReplace with the current isMock boolean. It could also fail because of LoadingIndicator.show() and .dispose are being called in a test. To avoid this you should surround them with an if statement to not run when isMock is true.

Commit:
Fix failing test

Ali-Alkhars committed:
Re-add features after merging master

- Re-add getItemAvailability
- Fix delete from everywhere
- Make sure that loading indictors show/hide only when not testing

Ali-Alkhars committed:
Merge branch 'master' into T19-Existing-Items

Ali-Alkhars committed:
Remove unused imports

Ali-Alkhars committed:
Add missing tests for edit category and delete category

- Add missing keys
- Add missing mock functions

Ali-Alkhars committed:
Add missing tests for delete item button

- Tests for delete item
- Tests for delete item everywhere

dmunipi1 reviewed (approved):
Good job guys
